task-maker-async/src/worker.rs

```diff
@@ -0,0 +1,566 @@
+                match path {
+                    ExecutionPath::Stdout => {
+                        Ok(FileSetFile::AuxiliaryFile(name, ExecutionFile::Stdout))
+                    }
+                    ExecutionPath::Stderr => {
+                        Ok(FileSetFile::AuxiliaryFile(name, ExecutionFile::Stderr))
+                    }
+                    ExecutionPath::Path(path) => {
+                        Ok(FileSetFile::AuxiliaryFile(name, ExecutionFile::File(path)))
+                    }
+                    ExecutionPath::Stdin => {
+                        bail!("Invalid file combination: Output + Stdin");
+                    }
+                }
+            };
+            match file {
+                ExecutionFileMode::Input(_) => {}
+                ExecutionFileMode::Fifo(_) => {}
+                ExecutionFileMode::Output => {
+                    let store_path = get_store_path(path)?;
+                    send_file_to_store(store, handle, store_path, local_path).await?
+                }
+            };
+            Result::<(), Error>::Ok(())
+        }
+    }))
+    .await?;
+    Ok(())
+}
+
+pub async fn evaluate_one_job(
+    id: usize,
+    server: ServerClient,
+    store: StoreClient,
+    sandbox_path: &Path,
+) -> Result<(), Error> {
+    let (execution_group, options, handle) = server.get_work(context::current(), id).await?;
+
+    // Immediately activate the handle for writing, creating a future that will resolve
+    // with the return value of activate_for_writing when file creation either finishes or the
+    // request is dropped.
+    let store_clone = store.clone();
+    let mut finished = tokio::spawn(async move {
+        store_clone
+            .activate_for_writing(context::current(), handle)
+            .await??;
+        Result::<(), Error>::Ok(())
+    });
+
+    let group_tempdir = TempDir::new_in(sandbox_path, "tm-execgroup")?;
+
+    let group_dir = if options.keep_sandboxes {
+        group_tempdir.into_path()
+    } else {
+        group_tempdir.path().into()
+    };
+
+    let fifo_dir = group_dir.join(FIFO_DIR_NAME);
+
+    create_dir_all(&fifo_dir).await?;
+
+    let all_fifos: HashSet<_> = execution_group
+        .executions
+        .iter()
+        .flat_map(|exec| exec.files.iter())
+        .filter_map(|(_, mode)| match mode {
+            crate::dag::ExecutionFileMode::Fifo(p) => Some(p),
+            _ => None,
+        })
+        .collect();
+
+    for fifo in all_fifos {
+        nix::unistd::mkfifo(&fifo_dir.join(fifo), nix::sys::stat::Mode::S_IRWXU)?;
+    }
+
+    let all_execs =
+        try_join_all(
+            execution_group.executions.iter().cloned().map(|execution| {
+                prepare_one_execution(store.clone(), execution, group_dir.clone())
+            }),
+        )
+        .await?;
+
+    let tool_path = task_maker_exec::find_tools::find_tools_path();
+
+    let run = try_join_all(
+        all_execs
```

Comment:
Does tokio guarantees that all the ready-to-be-started futures will be started, regardless of the number of waiting tasks?

Reply:
Why does that matter?